src/gui/codeeditors/qgscodeeditor.cpp

```diff
@@ -542,6 +589,31 @@ void QgsCodeEditor::populateContextMenu( QMenu * )
 
 }
 
+QString QgsCodeEditor::reformatCodeString( const QString &string )
+{
+  return string;
+}
+
+void QgsCodeEditor::showMessage( const QString &title, const QString &message, Qgis::MessageLevel level )
+{
+  switch ( level )
+  {
+    case Qgis::Info:
+    case Qgis::Success:
+    case Qgis::NoLevel:
+      QMessageBox::information( this, title, message );
+      break;
+
+    case Qgis::Warning:
+      QMessageBox::warning( this, title, message );
+      break;
+
+    case Qgis::Critical:
+      QMessageBox::critical( this, title, message );
+      break;
```

Comment:
A `QMessageBox `seems a bit much imho. Maybe we could add a `QgsMessageBar` for every `QgsCodeEditor`s (whose mode=`Mode::ScriptEditor`), kind of like the one in console_editor.py?

https://github.com/qgis/QGIS/blob/4c74bd99c8f46ca5c034532b17bc03ad993d13ea/python/console/console_editor.py#L635-L639

Reply:
It's tricky 😁. I've made the show message method virtual so that there is a way to redirect it to a message bar in subclasses (ie in the python console). But the QgsCodeEditor class directly inherits the Qscintilla base class, so we can't now change that to a more flexible Qwidget base without breaking API (and a Qwidget base would be needed to have extra widgets in a layout here). It's frustrating, because I'd really like to move more functionality to QgsCodeEditor like that search/replace bar, but that all requires the flexibility to modify the widget layout...